streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/KTableKTableForeignKeyJoinNode.java

```diff
@@ -0,0 +1,165 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.streams.kstream.internals.graph;
+
+import org.apache.kafka.common.serialization.Serde;
+import org.apache.kafka.common.utils.Bytes;
+import org.apache.kafka.streams.kstream.internals.Change;
+import org.apache.kafka.streams.kstream.internals.KeyValueStoreMaterializer;
+import org.apache.kafka.streams.kstream.internals.MaterializedInternal;
+import org.apache.kafka.streams.kstream.internals.foreignkeyjoin.CombinedKey;
+import org.apache.kafka.streams.kstream.internals.foreignkeyjoin.CombinedKeyByForeignKeyPartitioner;
+import org.apache.kafka.streams.processor.FailOnInvalidTimestamp;
+import org.apache.kafka.streams.processor.internals.InternalTopologyBuilder;
+import org.apache.kafka.streams.state.KeyValueStore;
+import org.apache.kafka.streams.state.StoreBuilder;
+
+/**
+ * Too much specific information to generalize so the Foreign Key KTable-KTable join requires a specific node.
+ */
+public class KTableKTableForeignKeyJoinNode<VR, K, V, KO, VO> extends StreamsGraphNode {
+    private final String repartitionTopicName;
+    private final String repartitionSinkName;
+    private final String repartitionSourceName;
+    private final ProcessorParameters repartitionProcessorParameters;
+    private final ProcessorParameters joinOneToOneProcessorParameters;
+    private final Serde<CombinedKey<KO, K>> combinedKeySerde;
+    private final Serde<V> thisValueSerde;
+    private final CombinedKeyByForeignKeyPartitioner<KO, K, V> partitioner;
+    private final MaterializedInternal<CombinedKey<KO, K>, V, KeyValueStore<Bytes, byte[]>> repartitionedPrefixScannableStore;
+    private final ProcessorParameters<KO, Change<VO>> joinByPrefixProcessorParameters;
+    private final String finalRepartitionTopicName;
+    private final String finalRepartitionSinkName;
+    private final String finalRepartitionSourceName;
+    private final Serde<K> thisKeySerde;
+    private final Serde<VR> joinedValueSerde;
+    private final ProcessorParameters<K, VR> highwaterProcessorParameters;
+    private final MaterializedInternal<K, Long, KeyValueStore<Bytes, byte[]>> highwaterMatInternal;
+    private final String finalRepartitionTableName;
+    private final String prefixScannableDBRefName;
+    private final MaterializedInternal<K, VR, KeyValueStore<Bytes, byte[]>> materialized;
+    private final ProcessorParameters<K, VR> outputProcessorParameters;
+    private final String thisName;
+    private final String otherName;
+    private final String[] otherValueGetterStoreNames;
+
+
+    //TODO - Can reduce some of the parameters, but < 13 is not possible at the moment.
+    //Would likely need to split into two graphNodes - ie: foreignKeyJoinNode and foreignKeyJoinOrderResolutionNode.
```

Comment:
I think the step here an optimizer _could potentially_ exploit is the repartitioning.

Reply:
I'll have to look more into the optimizer. TBH I built this originally in 1.0 and just did a functional port, not necessarily a best practices one. Thanks

Reply:
I would not recommend to spend to much energy. At the moment I really don't expect the optimizer to be able to exploit any of this. Probably also not in the future. Was just a though popping into my head